DROOLS-3826: [DMN Designer] Included Models - Decision Components - Create the component that lists all imported nodes

karreiro opened the pull request:
See:
- https://issues.jboss.org/browse/DROOLS-3826
- https://issues.jboss.org/browse/DROOLS-3828

![2019-04-10 22 09 48](https://user-images.githubusercontent.com/1079279/55924113-e129ef80-5bde-11e9-9bca-5e6ac3cfe455.gif)

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/included/components/DecisionComponentsItemView.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Override
+    public void setIcon(final String iconURI) {
+        icon.src = iconURI;
```
Just for consistency with other setters we could use `this.icon.src`

karreiro replied:
Generally I only use `this` to avoid ambiguity, and here `this` is not "required". So, keeping it as is 👀

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/included/components/DecisionComponentsView.java:
```diff
@@ -0,0 +1,161 @@
+
+    @PostConstruct
+    public void init() {
+        getDrgElementFilter().selectpicker("refresh");
```
In our codebase we use often `selectpicker("refresh")`, maybe we should introduce `refresh()` directly in `JQuerySelectPicker`?

karreiro replied:
I think we should keep the `JQuerySelectPicker` only for make the native implementation static typed. A good refactoring would be to extract this frequently used methods to a generic new `SelectPickerComponent`. But I'll postpone this refactoring for now, and in the next `selectpicker` I think we can considering the refactoring ☺️

jomarko commented:
@karreiro Also I would like to increase the coverage slightly https://github.com/karreiro/kie-wb-common/pull/6

jomarko commented:
### Manual review note 01
Saving a diagram with imported model throws an error:
```
09:51:18,571 ERROR [org.kie.dmn.core.compiler.DMNCompilerImpl] (EJB default - 9) Required import not found: Could not locate required dependency while importing DMN for namespace: https://github.com/kiegroup/drools/kie-dmn/_DA0D0D59-7965-40DD-BCE1-638DB8C4DE68, name: t, modelName: null. for node 'org.kie.dmn.model.v1_2.TImport@4e2d8e7b' 

```

manstis commented:
@jomarko The compilation issue is covered by https://issues.jboss.org/browse/DROOLS-3841

jomarko commented:
### Manual review note 02
- I am in state when I included 3 models
- All nodes shown [1]
- Then I filter by **All Included components** [2], here first issue appears that no all nodes are shown
- Then I am unable to select **Filter by** [3] as it was in the original state [1]

[1]
![Screenshot from 2019-04-11 10-11-20](https://user-images.githubusercontent.com/8044780/55942059-e1100b00-5c43-11e9-8040-c732721bc5fb.png)

[2]
![Screenshot from 2019-04-11 10-11-43](https://user-images.githubusercontent.com/8044780/55942071-e53c2880-5c43-11e9-81e1-96610867c49c.png)


[3]
![Screenshot from 2019-04-11 10-12-15](https://user-images.githubusercontent.com/8044780/55942079-ea997300-5c43-11e9-9b02-495c03d2e8e4.png)

jomarko commented:
### Manaul review note 03
I uploaded 4 dmn models into [DROOLS-3826](https://issues.jboss.org/browse/DROOLS-3826). The issue can be seen if **parent.dmn** is opened and user just change the node type.

See the [video](https://drive.google.com/file/d/1DYcE1yTDzXVyklYn4l8BiD4jbiE-VufW/view?usp=sharing).

jomarko commented:
@karreiro I think I am done with my review.

manstis commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/DMNIncludeModelsServiceImpl.java:
```diff
@@ -73,6 +81,16 @@ public DMNIncludeModelsServiceImpl(final RefactoringQueryServiceImpl refactoring
+                .stream()
+                .map(path -> includedNodesFilter.getNodesByNamespaces(path, namespaces))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
```
Personally I find the form shown by @karreiro easier to read.. but it's his code; he can decide his preference :-)

jomarko replied:
I agree, not crucial comment, I am fine if Guilherme won't change.

karreiro replied:
I agree with Mike, @jomarko. So, keeping it as is.. :eyes:

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/editors/types/DMNIncludeModelsService.java:
```diff
@@ -25,4 +25,7 @@
 
     List<DMNIncludeModel> loadModels(final WorkspaceProject rootPath);
+
+    List<DMNIncludedNode> loadNodesByNamespaces(final WorkspaceProject workspaceProject,
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/DMNIncludeModelsServiceImpl.java:
```diff
@@ -16,6 +16,7 @@
 
 package org.kie.workbench.common.dmn.backend.editors.types;
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-api/src/main/java/org/kie/workbench/common/dmn/api/editors/types/DMNIncludedNode.java:
```diff
@@ -0,0 +1,59 @@
+ * limitations under the License.
+ */
+
+package org.kie.workbench.common.dmn.api.editors.types;
```
Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/exceptions/DMNIncludedModelPathCouldNotBeRead.java:
```diff
@@ -0,0 +1,21 @@
+
+package org.kie.workbench.common.dmn.backend.editors.types.exceptions;
+
+public class DMNIncludedModelPathCouldNotBeRead extends Exception {
```
Removed!

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/included/components/DecisionComponentFilter.java:
```diff
@@ -0,0 +1,80 @@
+        return getOptionalString(term);
+    }
+
+    Stream<DecisionComponentsItem> query(Stream<DecisionComponentsItem> stream) {
```
😧 - I really hate missing finals. Thanks!!

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/decision/included/components/DecisionComponentFilterTest.java:
```diff
@@ -0,0 +1,160 @@
+    }
+
+    @Test
+    public void testGetTermWhenDrgElementIsNotPresent() {
```
Good catch! Fixed.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/included/components/DecisionComponentFilter.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    Stream<DecisionComponentsItem> query(Stream<DecisionComponentsItem> stream) {
+        return stream.filter(byDrgElement()).filter(byTerm());
```
I prefer to keep this just for filtering. But I'm sorting the list in the `DecisionComponents` class.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/included/components/DecisionComponents.java:
```diff
@@ -0,0 +1,183 @@
+
+    private final List<DecisionComponentsItem> decisionComponentsItems = new ArrayList<>();
+
+    private final DecisionComponentFilter filter = new DecisionComponentFilter();
```
Hmmm.. interesting. Done.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/decision/included/components/DecisionComponents.java:
```diff
@@ -0,0 +1,183 @@
+    public void refresh(final Diagram diagram) {
+
+        clearDecisionComponents();
+        enableLoading();
```
Fixed!

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/types/common/DMNDiagramHelper.java:
```diff
@@ -0,0 +1,84 @@
+        return getNamespace(diagram);
+    }
+
+    public String getNamespace(final Diagram diagram) {
```
Very good proposal. Thank you, @manstis :-)

karreiro commented:
Thank you for the review @jomarko and @manstis. Changes applied.

---

I've incorporated two extra things:

**1 - commit: 695532f6e2c08a74300dae093398a14bbb456953**
I've added the `getImportedElementId`. Daniel will need the DRGElement ID with the import name as the prefix - something like this: `model:0000-1111-2222-4444`. Now the `DMNIncludedNode` has this `getImportedElementId` method.

**2 - commit: b0d90a9e28885852206b1036ccb6434b2f03b51d**
Now a new import adds an entry in the `definitions.nsContext` (this is a bug fix).

---

@manstis, I've applied the suggested refactoring. Thank-you :-)

---

@jomarko,

I've applied most of your suggestions. I've left a note in each comment.

Regarding manual reviews..


**Manual review note 01**

Yeah, as @manstis mentioned this issue will be addressed by https://issues.jboss.org/browse/DROOLS-3841.
I've quickly debugged it, and it's not that simple. When a DMN file is saved, the project is automatically built (by using the incremental approach). I've noticed that in this build, external resources are not being loaded.


**Manual review note 02**

The _"Filter by"_ is not selectable, it's just the placeholder (the selectpicker library suggests the use of a disabled option as the placeholder).


**Manual review note 03**

I've downloaded the assets you provided, watched the video, and I could not reproduce the issue after the refactorings. Could you please double-check?

Thank-you.
